main/src/library/Iterator.flix

```diff
@@ -195,108 +174,84 @@ namespace Iterator {
     ///
     /// Consumes the entire iterator.
     ///
-    pub def toChain(iter: Iterator[a, r]): Chain[a] \ Read(r) =
+    pub def toChain(iter: Iterator[a, ef, r]): Chain[a] \ { ef, Read(r) } =
         foldLeft((acc, a) -> Chain.snoc(acc, a), Chain.empty(), iter)
 
     ///
     /// Returns the contents of `iter` as a map.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def toMap(iter: Iterator[(a, b), r]): Map[a, b] \ Read(r) with Order[a] =
-        foldLeft((acc, ab) -> Map.insert(fst(ab), snd(ab), acc), Map.empty(), iter)
+    pub def toMap(iter: Iterator[(k, v), ef, r]): Map[k, v] \ { ef, Read(r) } with Order[k] =
+        foldLeft((acc, a) -> {let (k, v) = a; Map.insert(k, v, acc)}, Map.empty(), iter)
 
     ///
     /// Returns the contents of `iter` as a `Some(Nel)` if `iter` is not empty. Returns None otherwise.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def toNel(iter: Iterator[a, r]): Option[Nel[a]] \ Read(r) = match toList(iter) {
-        case Nil     => None
-        case x :: xs => Some(Nel.Nel(x, xs))
-    }
+    pub def toNel(iter: Iterator[a, ef, r]): Option[Nel[a]] \ { ef, Read(r) } =
+        let swapCons = (a, acc) -> match acc {
+            case Nel(x, xs) => Nel(x, a :: xs)
+        };
+        match next(iter) {
+            case Some(a) => foldRight((x, acc) -> swapCons(x, acc), Nel.singleton(a), iter) |> Some
+            case None    => None
+        }
 
     ///
     /// Returns the contents of `iter` as a nec.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def toNec(iter: Iterator[a, r]): Option[Nec[a]] \ Read(r) =
-        let f = (x, acc) -> match acc {
-            case None      => Some(Nec.singleton(x))
-            case Some(nec) => Some(Nec.cons(x, nec))
-        };
-        foldRight(f, None, iter)
+    pub def toNec(iter: Iterator[a, ef, r]): Option[Nec[a]] \ { ef, Read(r) } = match next(iter) {
```

Comment:
What's the structure of the chain? I am trying to visualize it. I think it should like a list right? Which is probably exactly what you want?

Reply:
i.e. completely right leaning.

Reply:
I think you could do `toNec` as a `foldRight` but then I think you would need to make the accumulator an `Option` and have to test it at each step unless `foldRightWithCont` helps...